src/app/register/ProgramSelector.tsx

```diff
@@ -0,0 +1,236 @@
+import { cn, filterRedundantPrograms, isContainingName } from "@/lib/utils";
+import { Add, Close, Error, School, Warning } from "@mui/icons-material";
+import {
+  Autocomplete,
+  CircularProgress,
+  Option,
+  Select,
+  Tooltip,
+} from "@mui/joy";
+import { ProgramType } from "@prisma/client";
+import { useEffect, useMemo, useState } from "react";
+
+import { usePrograms } from "../hooks/reactQuery";
+import { Program } from "../types/Program";
+
+const MAX_PROGRAM_SELECTIONS = 2;
+
+interface ProgramSelectorProps {
+  programType: ProgramType;
+  selectedPrograms: Program[] | undefined;
+  addProgram: (program: Program) => void;
+  deleteProgram: (programId: number) => void;
+}
+
+export default function ProgramSelector({
+  programType,
+  selectedPrograms,
+  addProgram,
+  deleteProgram,
+}: ProgramSelectorProps) {
+  const [minorInput, setMinorInput] = useState("");
+  const [minorValue, setMinorValue] = useState<Program | null>(null); // undefined MUI components are uncontrolled
+  const [catalogYearInput, setCatalogYearInput] = useState("");
+
+  const [error, setError] = useState("");
+
+  // NOTE: Instead of fetching minors and years separately, fetch all programs
+  // and filter to reduce db calls
+  const { data: allPrograms, isPending, isFetching, isError } = usePrograms();
+  const programs = useMemo(() => {
+    const uniquePrograms = filterRedundantPrograms(allPrograms ?? []);
+    return uniquePrograms?.filter(
+      (program) => program.programType === programType,
+    );
+  }, [allPrograms, programType]);
+
+  const catalogYears = allPrograms?.filter(
+    (program) => program.name === (minorValue ? minorValue.name : ""),
+  );
+
+  const isMaxMajorsSelected = !!(
+    selectedPrograms && selectedPrograms.length >= MAX_PROGRAM_SELECTIONS
+  );
+
+  // Add a program to the list of selected programs
+  const handleAddProgram = () => {
+    setError("");
+
+    const programId = catalogYears!.find(
+      (program) => program.catalogYear === catalogYearInput,
+    )!.id;
+
+    if (
+      selectedPrograms &&
+      isContainingName(minorValue!.name, selectedPrograms)
+    ) {
+      setError("You have already added this major");
+      return;
+    }
+
+    if (isMaxMajorsSelected) return;
+
+    addProgram({
+      id: programId,
+      name: minorValue!.name,
+      catalogYear: catalogYearInput,
+      programType: programType,
+    });
+  };
+
+  // Delete a program from the list of selected programs
+  const handleDeleteProgram = (programId: number) => {
+    setError("");
+    deleteProgram(programId);
+  };
+
+  // NOTE: User thrown errors (more than one of same major) exist in addition to
```

Comment:
If this is the case, we might want to use either React Hook Form or Material/Joy UI forms

Reply:
I tried using React Hook form but realized the use case doesn't make sense for this particular case. The way error / warning messages are displayed requires hooking into the error state so makes sense just to control it manually.